internal/stacks/stackruntime/internal/stackeval/main.go

```diff
@@ -430,22 +428,25 @@ func (m *Main) RootVariableValue(ctx context.Context, addr stackaddrs.InputVaria
 		if !m.Applying() {
 			panic("using ApplyPhase input variable values when not configured for applying")
 		}
-		ret, ok := m.applying.rootInputVals[addr]
-		if !ok {
-			// We should not get here if the given plan was created from the
-			// given configuration, since we should always record a value
-			// for every declared root input variable in the plan.
+
+		// First, check the values given to use directly by the caller.
+		if ret, ok := m.applying.opts.InputVariableValues[addr]; ok {
+			return ret
+		}
+
+		// If the caller didn't provide a value, we need to look up the value
+		// that was used during planning.
+
+		if ret, ok := m.applying.plan.RootInputValues[addr]; ok {
 			return ExternalInputValue{
-				Value: cty.DynamicVal,
+				Value: ret,
 			}
 		}
-		return ExternalInputValue{
-			Value: ret,
 
-			// We don't save source location information for variable
-			// definitions in the plan, but that's okay because if we were
-			// going to report any errors for these values then we should've
-			// already done it during the plan phase, and so couldn't get here..
+		// This shouldn't happen because we should always have a value for
```

Comment:
If this shouldn't happen, can we panic, emit a diagnostic?

Reply:
So, I went in to make it panic. Then decided I wanted to see what happens now if I forced this, then realised that this is totally possible if the caller forgets to set an ephemeral value at apply time. So, I added a test for that, and updated this to return a null value which means the input value parsing logic will now fail with a nice error message. Updated the comment to reflect this.